tla-types/src/main/scala/at/forsyte/apalache/tla/typecheck/etc/ConstraintSolver.scala

```diff
@@ -77,6 +77,16 @@ class ConstraintSolver(approximateSolution: Substitution = Substitution.empty) {
     }
   }
 
+  /**
+   * Test whether a variable is free in the context that is induced by the solved constraints.
+   *
+   * @param varNo a variable number
+   * @return true if the variable occurs in the partial solution of the solver
+   */
+  def isFreeVar(varNo: Int): Boolean = {
+    approximateSolution.mapping.keySet.forall(cls => !cls.typeVars.contains(varNo))
+  }
+
```

Comment:
Any chance that we'll need to hit this frequently enough that it'll effect performance having to search through every item each time we check for free vars? If so, perhaps (as a followup?) we should keep a cash of free variables, and then just remove vars from the cache as they are bound?

Reply:
I hope there is no big drawback. This check is done only for a few variables per definition, and TLA+ specs are below 2-3 KLOC. There is another trick in Chapter 10 of Advanced Types. I did not get to it, but it may be a good idea to read about it :)

Reply:
In any case, @Kukovec  wanted to rewrite the unifier with union-find in #973. So this code will most likely change.

Reply:
Shall I bump that up in priority then?

Reply:
I think so.